lib/ansible/modules/windows/win_zip.ps1

```diff
@@ -0,0 +1,253 @@
+    {
+        PowerShell,
+        DotNet,
+        Undefined
+    }
+"@
+
+$try_dotnet = $false
+try {
+    # Determines if the Compress-Archive command is available (preferred use)
+    # if not will check if dotnet libs are available (.NET 4.5 or greater)
+    Get-Command -Name Compress-Archive | Out-Null
+    $utils_available = [CompressionUtils]::PowerShell
+}
+catch {
+    $try_dotnet = $true
+}
+
+if ($try_dotnet) {
+    try {
+        # determines if .net 4.5 is available, if this fails we can't
+        # fall back to the legacy COM Shell.Application to extract the zip because
+        # the options don't allow it to be suppressed in the UI
+        Add-Type -AssemblyName System.IO.Compression.FileSystem | Out-Null
+        Add-Type -AssemblyName System.IO.Compression | Out-Null
+        $utils_available = [CompressionUtils]::DotNet
+    }
+    catch {
+        # default to Undefined
+        $utils_available = [CompressionUtils]::Undefined
+    }
+
+}
+
+$result.utils_used = $utils_available.ToString()
+# END: Determine which type of compression utils to use: PowerShell or .NET or COM (to follow dev standards)
+
+if ($utils_available -eq [CompressionUtils]::Undefined) {
+    Fail-Json -obj $result -message "Could not find any available compression utilities: PowerShell 5.0 and .NET Framework 4.5 are minimally required.  Please install the Windows Management Framework 5.0 or higher."
+}
+
+# Validate that the extension of the destination zip file is 'zip' - only one supported right now
+$src_ext = (Split-Path $dest -Leaf).Split('.')[-1]
+if ($src_ext -ne "zip") {
+    Fail-Json -obj $result -message "The destination zip file specified '$dest' must have an extension of '.zip'."
+}
+
+# Validate src path exists
+# Ansible.ModuleUtils.Legacy.psm1 checks this for a type of 'PATH' so right now this is duplicate
+# but will help if the type is changed to 'LIST' to support single or multiple src values
+# Also may want to look at adding support here for expanding environment vars when moving to list
+# https://github.com/ansible/ansible/blob/v2.4.1.0-1/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Legacy.psm1#L209
+
+If (-not (Test-Path -LiteralPath $src)) {
+    Fail-Json -obj $result -message "The source path specified '$src' could not be found."
+}
+
+# Ensure dest zip DOES NOT exist unless creates attribute specified (TODO: support "FORCE" option)
+if ($creates -and (Test-Path -LiteralPath $creates)) {
+    Exit-Json -obj $result -message "The destination zip file '$dest' specified in the creates attribute exists.  Nothing changed"
+}
+elseif (Test-Path -LiteralPath $dest) {
+    Fail-Json -obj $result -message "The destination zip file specified '$dest' already exists. Adding files to existing zips not yet supported"
+}
+
+# Ensure we don't have any files over 2GB in size (PowerShell & .NET Limitation)
+$large_files = New-Object -Type System.Collections.ArrayList
+Find-LargeFiles -src $src -large_files ([ref]$large_files)
```

Comment:
You shouldn't be passing in an arraylist as an argument and having the function modify it, technically you don't even need to use [ref] as that type is already passed by ref.

What should happen is `Find-LargeFiles` should output all the files larger than the size specified and you store that in a variable, i.e.

```
$large_files = Find-LargeFiles -src $src

if ($large_files) {
    ...
}
```

Also .NET doesn't seem to have a problem with files over 2GB, I just tested `[System.IO.Compression.ZipFile]::CreateFromDirectory()` that contains a large file and it worked just fine.

Reply:
From the [PowerShell docs](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.archive/compress-archive?view=powershell-5.1#description)

> The Compress-Archive cmdlet uses the Microsoft .NET API System.IO.Compression.ZipArchive to compress files. The maximum file size is 2 GB because there's a limitation of the underlying API.

I tested and ran into this on my 2008/2012/2016 boxes. Not clear why you're not experiencing the same.

Reply:
I encountered the issue with `Compress-Archive` but not with the .NET call. Using `CreateFromDirectory()` here is a 5GB file in the archive

![image](https://user-images.githubusercontent.com/8462645/72848963-7a9b3b80-3cf1-11ea-8298-9db142c0ad31.png)

I even tested the other way of using `CreateEntryFromFile()` and that was also fine.

After looking into the docs a bit more I think I know where the confusion is coming from. The docs for [CreateEntryFromFile](https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.zipfileextensions.createentryfromfile?view=netframework-4.8) state at the bottom

> When ZipArchiveMode.Update is present, the size limit of an entry is limited to Int32.MaxValue. This limit is because update mode uses a MemoryStream internally to allow the seeking required when updating an archive, and MemoryStream has a maximum equal to the size of an int.

You aren't using the Update mode, just Open, and this somewhat lines up with your check as `[Int32]::MaxValue` is `2147483647`. It's a bit bigger than `2097152000` in your code but `2GB` is `2147483648` bytes so I'm not sure what the value in the module is from. There is also the WinRM memory limitation that will be encountered if a memory stream is being used to store the full archive but from what I can see the code you have today shouldn't be doing that.